packages/hydrogen/src/cart/optimistic/useOptimisticCart.tsx

```diff
@@ -0,0 +1,128 @@
+import {useFetchers} from '@remix-run/react';
+import type {cartGetDefault} from '../queries/cartGetDefault';
+import {CartForm} from '../CartForm';
+import type {
+  Cart,
+  CartLine,
+  ProductVariant,
+} from '@shopify/hydrogen-react/storefront-api-types';
+import {
+  getOptimisticLineId,
+  isOptimisticLineId,
+} from './optimistic-cart.helper';
+
+export type OptimisticCart<T = ReturnType<typeof cartGetDefault>> = T & {
+  isOptimistic?: boolean;
+  lines: {nodes: Array<CartLine & {isOptimistic?: boolean}>};
+};
+
+/**
+ * @param cart The cart object from `context.cart.get()` returned by a server loader.
+ *
+ * @returns A new cart object augmented with optimistic state. Each cart line item that is optimistically added includes an `isOptimistic` property. Also if the cart has _any_ optimistic state, a root property `isOptimistic` will be set to `true`.
+ */
+export function useOptimisticCart<
+  DefaultCart = ReturnType<typeof cartGetDefault>,
+>(cart: DefaultCart): OptimisticCart<DefaultCart> {
+  const fetchers = useFetchers();
+
+  if (!fetchers || !fetchers.length) return cart as OptimisticCart<DefaultCart>;
+
+  let isOptimistic = false;
+
+  const optimisticCart = cart
+    ? (structuredClone(cart) as OptimisticCart<DefaultCart>)
+    : ({lines: {nodes: []}} as OptimisticCart<DefaultCart>);
```

Comment:
It is possible that `cart={}`
not sure how, I re-product by linking and unlinking a shop, then add a product into a fresh cart

checking for cart.lines seems to be an ok fix